easy_rec/python/tools/feature_selection.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import division
+from __future__ import print_function
+
+import json
+import os
+from collections import OrderedDict
+
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+from easy_rec.python.utils import config_util
+
+matplotlib.use('Agg')
+
+tf.app.flags.DEFINE_string('model_type', 'variational_dropout',
+                           'feature selection model tyoe')
+tf.app.flags.DEFINE_string('config_path', '',
+                           'feature selection model config path')
+tf.app.flags.DEFINE_string('checkpoint_path', None,
+                           'feature selection model checkpoint path')
+tf.app.flags.DEFINE_string('output_dir', '',
+                           'feature selection result directory')
+tf.app.flags.DEFINE_integer('topk', 100, 'select topk importance features')
+tf.app.flags.DEFINE_string('fg_path', '', 'fg config path')
+tf.app.flags.DEFINE_bool('visualize', False,
+                         'visualization feature selection result or not')
+FLAGS = tf.app.flags.FLAGS
+
+
+class VariationalDropoutFS:
+
+  def __init__(self,
+               config_path,
+               output_dir,
+               topk,
+               checkpoint_path=None,
+               fg_path=None,
+               visualize=False):
+    self._config_path = config_path
+    self._output_dir = output_dir
+    self._topk = topk
+    if not tf.gfile.Exists(self._output_dir):
+      tf.gfile.MakeDirs(self._output_dir)
+    self._checkpoint_path = checkpoint_path
+    self._fg_path = fg_path
+    self._visualize = visualize
+
+  def process(self):
+    tf.logging.info('Loading logit_p of VariationalDropout layer ...')
+    feature_dim_dropout_p, embedding_wise_variational_dropout = self._feature_dim_dropout_ratio(
+    )
+
+    tf.logging.info('Calculating feature importance ...')
+    feature_importance = self._get_feature_importance(
+        feature_dim_dropout_p, embedding_wise_variational_dropout)
+
+    tf.logging.info('Processing model config ...')
+    self._process_config(feature_importance)
+
+    tf.logging.info('Dump feature importance to csv ...')
+    self._dump_to_csv(feature_importance)
+
+    if self._visualize:
+      tf.logging.info('Visualizing feature importance ...')
+      if embedding_wise_variational_dropout:
+        self._visualize_embedding_dim_importance(feature_dim_dropout_p)
+      self._visualize_feature_importance(feature_importance)
+
+  def _feature_dim_dropout_ratio(self):
+    """Get dropout ratio of embedding-wise or feature-wise."""
+    config = config_util.get_configs_from_pipeline_file(self._config_path)
+    embedding_wise_variational_dropout = config.model_config.variational_dropout.embedding_wise_variational_dropout
```

Comment:
assert config.model_config.HasField('variational_dropout'), 'variational_dropout must be in model_config'

Reply:
done